perfkitbenchmarker/scripts/object_storage_api_tests.py

```diff
@@ -117,6 +119,10 @@
 # A global variable initialized once for Azure Blob Service
 _AZURE_BLOB_SERVICE = None
 
+# boto3 recommends creating per-thread sessions. We store them in
+# THREAD_LOCAL.boto_session.
```

Comment:
Before you check in there final changes please add side by side comparison
numbers.  Before/after boto changes.
On Feb 25, 2016 5:54 PM, "noahl" <notifications@github.com> wrote:

> In perfkitbenchmarker/scripts/object_storage_api_tests.py
> <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/852#discussion_r54195964>
> :
>
> > @@ -117,6 +119,10 @@
> >  # A global variable initialized once for Azure Blob Service
> >  _AZURE_BLOB_SERVICE = None
> >
> > +# boto3 recommends creating per-thread sessions. We store them in
> > +# THREAD_LOCAL.boto_session.
>
> It took a lot of following code, but I learned that the botocore library
> (which boto3 is based on) uses Requests (python-requests.org) for its
> HTTP(S) functionality, and Requests supports connection pooling.
>
> I know there is one pool per boto3.Endpoint object. I *think* that
> different threads have different Endpoints, so we have a connection pool
> per thread. It would be pretty difficult to override this - we would need
> to make our own botocore.Client objects, which is a few layers of
> abstraction below where this PR is. The alternative is to use the
> workaround in boto/botocore#577
> <https://github.com/boto/botocore/issues/577> to share one session object
> across threads. Unfortunately, I noticed that the connection pool holds a
> maximum of 10 connections, so that might be a problem if we had lots of
> threads.
>
> Do you think it would work for different threads to have their own
> connections? We're not doing much computation, so I think threads are
> spending most of their time on I/O anyway, so they're probably using their
> connection objects pretty well. I think that we would only experience
> overhead from having too many connections if we got to the point where we
> were saturating our network bandwidth, and in that case we would also have
> overhead from too many threads. So I think it might be okay to have
> per-thread connections. How does that sound to you?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/852/files#r54195964>
> .
>